canonical/vanilla-framework
# Show Jinja Macro usage in pattern code snippets

## Done

- Adds option for examples that are intended to be used as macros to be rendered as Jinja templates in example code snippets. Users have the option to switch between Jinja & HTML.
  - To show the Jinja template of an example in the documentation, add `data-lang="jinja"` to the placement element in the docs.
  - Refactored example Flask routes to unify their rendering logic under a single `example()` function. Added a route for `/docs/example/<example_file_path>?raw=true` that serves raw template files. This was necessary to expose the Jinja templates of examples to the front-end so they can be rendered in docs page code snippets.
- Drive-by:
  - Some improved documentation of the `example.js` file
  - Change all usages of `var` in `example.js` to `let` or `const`
  - Moved non-example usages of `throttle()` to a shared utils script
  - Changed the remote fetching protocol for example snippets from `XMLHttpRequest` to the more modern `fetch`. Extracted the remote fetching logic into its own `async` function, for simpler handling of multiple network requests.
  - Formatted all code sent to CodePen with `js-beautify` (fixes #5332)

Fixes #5254  
Fixes #5332 
Fixes [WD-13705](https://warthogs.atlassian.net/browse/WD-13705)

## QA

- Test every currently supported state of example code snippet and make sure it works as expected.
  - HTML only: Should have no language selector and display HTML. Ex: [CTA Block](https://vanilla-framework-5321.demos.haus/docs/patterns/cta-block)
  - Multiple languages, built without macros: Should have language selector with an option for each language, & default to HTML. Ex: [Accordion](https://vanilla-framework-5321.demos.haus/docs/patterns/accordion#default)
  - Multiple languages, built with macros (that we don't want to display): Same as previous. Ex: [Reduced navigation](https://vanilla-framework-5321.demos.haus/docs/patterns/navigation#reduced)
  - HTML built with Jinja macros (that we want to display): Should default to HTML, but allow switching between Jinja & HTML. Ex: [Hero 50/50](https://vanilla-framework-5321.demos.haus/docs/patterns/hero#5050)
- Verify that all examples that should show macro calls have the "Jinja" option in their code snippets and default to HTML. Verify you can also switch to Jinja on those examples.
  - Hero
  - Tiered List
- Open any standalone example and verify that is uses the standalone stylesheet. For example, try [button](https://vanilla-framework-5321.demos.haus/docs/examples/standalone/base/button?theme=light). You can verify the standalone stylesheet is used by opening the browser network tools and verifying that `build.css` is not loaded on a hard-refresh.
- Find an example that uses multiple languages (such as [reduced navigation](https://vanilla-framework-5321.demos.haus/docs/patterns/navigation#reduced)) and open the example in CodePen. Verify that all languages are populated in CodePen, and all languages are indent-formatted.

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)


## Screenshots
<img width="979" alt="Screenshot 2024-08-22 at 2 31 36 PM" src="https://github.com/user-attachments/assets/09cd920c-e766-4024-bee7-476e2105b279">
<img width="979" alt="Screenshot 2024-08-22 at 2 31 45 PM" src="https://github.com/user-attachments/assets/9f36ff6f-5954-4564-8cbe-03141afa5534">



[WD-14490]: https://warthogs.atlassian.net/browse/WD-14490?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[WD-13705]: https://warthogs.atlassian.net/browse/WD-13705?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
